modules/file-transfer/src/main/java/org/apache/ignite/internal/network/file/FileTransferServiceImpl.java

```diff
@@ -0,0 +1,402 @@
+
+    @Override
+    public CompletableFuture<Void> upload(String nodeConsistentId, Metadata metadata) {
+        return supplyAsync(() -> getFileProvider(metadata))
+                .thenCompose(provider -> provider.files(metadata))
+                .thenCompose(files -> {
+                    FileUploadRequest message = factory.fileUploadRequest()
+                            .metadata(metadata)
+                            .build();
+
+                    if (files.isEmpty()) {
+                        return failedFuture(new FileTransferException("No files to upload"));
+                    } else {
+                        return messagingService.invoke(nodeConsistentId, FILE_TRANSFERRING_CHANNEL, message, RESPONSE_TIMEOUT)
+                                .thenApply(FileUploadResponse.class::cast)
+                                .thenCompose(response -> sendFiles(nodeConsistentId, response.transferId(), files));
+                    }
+                });
+    }
+
+    private static short getMessageType(Class<?> metadata) {
+        Transferable annotation = metadata.getAnnotation(Transferable.class);
+        if (annotation == null) {
+            throw new IllegalArgumentException("Class " + metadata.getName() + " is not annotated with @Transferable");
+        }
+        return annotation.value();
+    }
+
+    private <M extends Metadata> FileProvider<M> getFileProvider(M metadata) {
+        FileProvider<Metadata> provider = metadataToProvider.get(metadata.messageType());
+        if (provider == null) {
+            throw new FileProviderNotFoundException(metadata.getClass());
+        } else {
+            return (FileProvider<M>) provider;
+        }
+    }
+
+    private <M extends Metadata> FileConsumer<M> getFileHandler(M metadata) {
+        FileConsumer<Metadata> handler = metadataToHandler.get(metadata.messageType());
+        if (handler == null) {
+            throw new FileHandlerNotFoundException(metadata.getClass());
+        } else {
+            return (FileConsumer<M>) handler;
+        }
+    }
+
+    private Path createTransferDirectory(UUID transferId) {
+        try {
+            return Files.createDirectories(tempDirectory.resolve(transferId.toString()));
+        } catch (IOException e) {
+            throw new FileTransferException("Failed to create transfer directory. Transfer id: {}. Exception: {}" + transferId, e);
+        }
+    }
+
+    private static void deleteDirectoryIfExists(Path directory) {
+        try {
+            FilesUtils.deleteDirectoryIfExists(directory);
+        } catch (IOException e) {
+            LOG.error("Failed to delete directory: {}. Exception: {}", directory, e);
+        }
+    }
+
+    private static Throwable unwrapFileTransferException(Throwable throwable) {
+        Throwable unwrapped = ExceptionUtils.unwrapCause(throwable);
+        if (unwrapped instanceof FileTransferException) {
+            return unwrapped.getCause() == null ? unwrapped : unwrapped.getCause();
```

Comment:
Wow, why so complicated? I would love to have a comment here, or a simpler solution when it comes to exceptions